python/phase-1.md

```diff
@@ -0,0 +1,441 @@
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+
+async def main():
+    core = await temporalio.Core.default_local()
+    client = temporalio.Client(core)
+    result = await client.execute_workflow(
+        "MyWorkflow", "arg1", id="someid", task_queue="somequeue"
+    )
+    print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+
+def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+
+async def main():
+    core = await temporalio.Core.default_local()
+    worker = temporalio.Worker(
+        core, task_queue="somequeue", activities={"say_hello": say_hello}
+    )
+    await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Core
+
+Exported from `temporalio` (but probably in `temporalio/core/core.py`, exported from `temporalio.core` then
+exported from `temporalio`):
+
+```python
+class Core:
+    @staticmethod
+    async def create(
+        addr: str,
+        *,
+        namespace: str = "default"
+        # ... additional options omitted for brevity
+    ) -> Core:
+        pass
+
+    @classmethod
+    async def default_local(cls) -> Core:
+        pass
+
+    async def shutdown(self) -> None:
+        pass
+```
+
+Notes:
+
+* After discussion, it has been decided that both clients and workers will be backed by core
+    * Yes that means clients require the Rust library which we accept
+* After discussion, it has been decided that we will not have an implicit local default like TypeScript SDK, but instead
+  an explicit default
+  * `default_local()` memoizes success
+* Should we have a constructor, an `__aenter__`, and an `__aexit__`?
+  * Probably
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(core_or_service: temporalio.Core) -> None:
```

Comment:
```suggestion
    def __init__(core: temporalio.Core) -> None:
```

(leftover from an old version that accepted a union of core or pure Python gRPC which we don't want atm)

Reply:
Fixed.